File: Day 3/Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [1] * n

        # Left pass: ans[i] holds product of all elements to the left of i
        left = 1
        for i in range(n):
            ans[i] = left
            left *= nums[i]

        # Right pass: accumulate product of elements to the right and multiply in-place
        right = 1
        for i in range(n - 1, -1, -1):
            ans[i] *= right
            right *= nums[i]

        return ans

File: Day 3/Product of Array Except Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        int left = 1, right= 1;
        vector<int> ans(n);

        //add ones in vector ans
        for(int i = 0; i < n;i++){
            ans[i] = 1;
        }

        // Left pass: ans[i] holds product of all elements to the left of i
        for(int i = 0; i < n; i++){
            ans[i] *= left;
            left *= nums[i];
        }

        //Right pass: accumulate product of elements to the right and multiply in-place
        for(int i = n-1; i >= 0; i--){
            ans[i] *= right;
            right *= nums[i];
        }

        return ans;
    }
};
